src/python/grpcio_reflection/grpc_reflection/v1alpha/proto_reflection_descriptor_database.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2021 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Reference implementation for reflection client in gRPC Python."""
+
+from google.protobuf.descriptor_database import DescriptorDatabase
+from google.protobuf.descriptor_pb2 import FileDescriptorProto
+import grpc
+from grpc_reflection.v1alpha.reflection_pb2 import (
+    ErrorResponse,
+    ExtensionNumberResponse,
+    ExtensionRequest,
+    FileDescriptorResponse,
+    ListServiceResponse,
+    ServerReflectionRequest,
+    ServerReflectionResponse,
+    ServiceResponse,
+)
+from grpc_reflection.v1alpha.reflection_pb2_grpc import ServerReflectionStub
+import logging
+
+
+class ProtoReflectionDescriptorDatabase(DescriptorDatabase):
+    """
+    ProtoReflectionDescriptorDatabase takes a stub of ServerReflection and provides the methods defined by
+    DescriptorDatabase interfaces. It can be used to feed a DescriptorPool instance.
+
+    Python implementation based on C++ version found here:
+      https://github.com/grpc/grpc/blob/v1.39.1/test/cpp/util/proto_reflection_descriptor_database.cc
+    while implementing the interface given here:
+      https://googleapis.dev/python/protobuf/3.17.0/google/protobuf/descriptor_database.html
+
+    """
+
+    def __init__(self, channel: grpc.Channel):
+        DescriptorDatabase.__init__(self)
+        self._logger = logging.getLogger(__name__)
+        self._stub = ServerReflectionStub(channel)
+        self._known_files: Set[str] = set()
+        self._cached_extension_numbers: Dict[str, List[int]] = dict()
+        self._cached_extension_files: Dict[str, FileDescriptorProto] = dict()
+
+    def _DoOneRequest(
+        self, request: ServerReflectionRequest
+    ) -> ServerReflectionResponse:
+        response = self._stub.ServerReflectionInfo(iter([request]))
+        res = next(response)
+        if res.WhichOneof("message_response") == "error_response":
+            raise grpc.RpcError(res.error_response)
```

Comment:
Can I ask what would the error message look like?

Reply:
I looked further into this, and indeed this flow isn't correct (I assume that's where you were going with your question 😃).

Making a call such as `desc_pool.FindFileByName("foo")` first calls the C++ layer: [FindFileByName](https://github.com/protocolbuffers/protobuf/blob/63f952b987bd848ac278c714c81c47250214adc7/python/google/protobuf/pyext/descriptor_pool.cc#L263) -> PyDescriptorDatabase::FindFileByName, and in case of error it checks its type. If it's not a KeyError it [prints a stack trace](https://github.com/protocolbuffers/protobuf/blob/63f952b987bd848ac278c714c81c47250214adc7/python/google/protobuf/pyext/descriptor_database.cc#L64). Then it always triggers a KeyError with the message ["Couldn't find file ..."](https://github.com/protocolbuffers/protobuf/blob/63f952b987bd848ac278c714c81c47250214adc7/python/google/protobuf/pyext/descriptor_pool.cc#L235) - which IMO is not the correct message, since usually the value of a KeyError is the key that was searched, but that's out of scope.

So the conclusion is that here in Python code we should verify that the error is a NOT_FOUND error. If it is, then raise a KeyError (which sadly will be transformed to another KeyError by the C++ layer). Otherwise, it's just an AssertionError, since we shouldn't get any other error here anyway (and again, sadly this AssertionError will again be converted to a KeyError, but in addition will get logging of the original error).

I'll also add this to the unit tests.